cdmtaskservice/dockerimages.py

```diff
@@ -2,29 +2,151 @@
-        # TODO throw an exception if something goes wrong
-        return "sha256:fake"
+        image_name = _validate_image_name(image_name)
+        retcode, stdo, stde = await self._run_crane("digest", image_name)
+        if retcode > 0:
+            # TODO LOGGING figure out how this is going to work, want to associate logs with
+            #              user names, ips, etc.
+            # TDOO TEST logging add tests for logging before exiting prototype stage
+            #           manual testing for now
+            logging.getLogger(__name__).error(
+                f"crane lookup of image {image_name} failed, retcode: {retcode}, stderr:\n{stde}")
+            # special case when crane spits out html and the regular error parsing doesn't work
+            if "404 not found" in stde.lower():
+                raise DigestFetchError(
+                    f"Failed to get digest for image {image_name}. "
+                    + "Image was not found on the host")
+            # This is pretty fragile for non-docker repository hosts.
+            # Not really sure what else to do here
+            # Allowlist for hosts?
+            errcode = stde.split("\n")[-1].split(':')[-1].strip()
+            raise DigestFetchError(
+                f"Failed to get digest for image {image_name}. Error code was: {errcode}")
+        return stdo
+
+    async def _run_crane(self, *args) -> (int, str, str):
+        # crane has very small output so just buffer in memory
+        try:
+            proc = await asyncio.create_subprocess_exec(
+                self._crane, *args,
+                stdout=asyncio.subprocess.PIPE,
+                stderr=asyncio.subprocess.PIPE
+            )
+        except FileNotFoundError as e:
+            raise CranePathError(
+                f"Configured crane executable path {self._crane} is invalid") from e
+        stdo, stde = await proc.communicate()
+        return proc.returncode, stdo.decode().strip(), stde.decode().strip()
+
+
+def _validate_image_name(image_name: str) -> str:
+    image_name = image_name.strip() if image_name is not None else None
+    # Image name rules: https://docs.docker.com/reference/cli/docker/image/tag/
+    # Don't do an exhaustive check here, but enough that we're reasonably confident
+    # it's a real name and there's no shell injection risk
+    # Also allows us to provide more specific error messages
+    if not image_name:
+        raise ImageNameParseError("No image name provided")
+    parts = image_name.split("/")
+    # assume a namespace is always present
+    if len(parts) == 2:  # docker
+        host, repo, image_and_tag = None, parts[0], parts[1]
+    elif len(parts) == 3:
+        host, repo, image_and_tag = parts
+    else:
+        # theoretically can have > 2 slashes, but don't bother with that case for now.
+        raise ImageNameParseError(f"Expected 1 or 2 '/' symbols in image name '{image_name}'")
```

Comment:
Why won't images like `python:3.9-slim` be considered a valid image name for us?

Reply:
Hmm, good catch, although I doubt we'll ever use an image like that to run a cdm job, so maybe it's ok. What do you think?

Reply:
I think that's okay. Maybe just add a note so that we know our limitations.

Reply:
What's interesting is if you have something like `myrepo.io/myimage:tag` how do you know whether `myrepo.io` is a namespace or a tag?

Reply:
Updated the pydoc for the method

Reply:
👍